proposals/p5913.md

```diff
@@ -0,0 +1,938 @@
+implementation, use it to test for unformed state, and adjust the operation to
+initialize (rather than assign) or skip destruction.
+
+Unformed objects can be directly created from the unformed initialization
+struct:
+
+```carbon
+impl forall [T:! UnformedInit]
+    T.(UnformedInit.StructT)
+    as ImplicitAs(Core.MaybeUnformed(T)) {
+  fn Convert[self: Self]() -> Core.MaybeUnformed(T) {
+    // Initialize the fields of `T` that are part of `StructT`.
+  }
+}
+```
+
+We also allow types to document that they have checked the type's invariants
+outside of what the typesystem can prove by using the unsafe as operation. For
+example:
+
+1. Convert a fully initialized `Core.MaybeUnformed(T)` object back to `T`.
+2. Convert an unformed `Core.MaybeUnformed(T)` object to its underlying raw
+   storage.
+
+Both of these convert _reference expressions_, and continue to refer to the same
+underlying storage or object. Because these are all _compatible_, they can also
+be used to convert pointers between these types.
+
+The first is useful when the `Core.MaybeUnformed(T)` object is in fact fully
+initialized to continue handling that case. It can also be used when the
+unformed state is arranged to be a valid initialization of the object and the
+operation is merely reifying it as fully initialized.
+
+The second is useful when the `Core.MaybeUnformed(T)` object is in fact in an
+unformed state and a new object will be initialized into its existing raw
+storage (because the destructor need not be run for such an object). The
+language mechanisms needed for raw storage and initializing it this will be part
+of specifying raw storage in a future proposal.
+
+**Future work:** we should consider adding `impl`s to `Core.MaybeUnformed(T)`
+when `IsUnformed` is implemented, ideally matching the those used for optional
+types. This should allow `Core.MaybeUnformed(T)` to participate in all the
+language-level affordances we provide for optional types as they are designed. A
+key goal should be that this allows using `Core.MaybeUnformed(T)` without any
+unsafe operations through tools like if let.
+
+## Flow-sensitive restrictions on objects known to be unformed
+
+Finally, we tie the previous two components together with flow-sensitive
+restrictions on how objects known to be unformed can be used.
+
+An object is known to be in an unformed state when it is declared without an
+explicit initializer, such as:
+
+```carbon
+var object: SomeType;
+```
+
+**Future work:** We also expect to add operations to Carbon that put objects
+into this state without a declaration, as we would like to use unformed state
+for non-destructively-moved-from objects as well, but those will come in
+subsequent and separate proposals.
+
+Because object above is known to be in an unformed state, we restrict how it may
+be used subsequently: it be used freely with the special type
+`Core.MaybeUnformed(T)`, but may only be used as the type `T` after it has been
```

Comment:
How does one use `object` with this special type? What's the syntax for that? (Do you just name it -- requiring us to do the control flow analysis online?)